File: java/src/main/java/S021.java
import java.util.stream.IntStream;

public class S021 {

    public static final int LIMIT = 10 * 1000;

    public static int solve() {
        return IntStream.range(2, LIMIT).filter(S021::isAmicable).sum();
    }

    private static boolean isAmicable(int n) {
        int partner = sumProperDivisors(n);
        return partner != n && sumProperDivisors(partner) == n;
    }

    private static int sumProperDivisors(int n) {
        return Maths.getDivisors(n).stream().mapToInt(Long::intValue).sum() - n;
    }

}

File: cpp/S021.cpp
#include <iostream>
#include "maths.cpp"

using namespace std;

static const int LIMIT = 10 * 1000;

int proper_divisor_sum(int n) {
    return sum_divisors_brute_force(n) - n;
}

bool is_amicable_sum(int n) {
    int pds = proper_divisor_sum(n);
    if (pds == n) {
        return false;
    } else {
        return proper_divisor_sum(pds) == n;
    }
}

int main() {
    int sum = 0;
    for (int i = 2; i < LIMIT; i++) {
        sum += i * is_amicable_sum(i);
    }
    cout << sum << endl;
    return 0;
}
